File: Android/app/src/main/java/com/dreamproject/inwords/data/repository/translation/TranslationWordsWebApiRepository.java
package com.dreamproject.inwords.data.repository.translation;

import com.dreamproject.inwords.data.dto.EntityIdentificator;
import com.dreamproject.inwords.data.dto.WordTranslation;
import com.dreamproject.inwords.data.source.webService.WebRequestsManager;
import com.dreamproject.inwords.data.sync.PullWordsAnswer;
import com.dreamproject.inwords.domain.util.WordsUtilKt;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

public class TranslationWordsWebApiRepository implements TranslationWordsRemoteRepository {
    private WebRequestsManager webRequestsManager;

    @Inject
    public TranslationWordsWebApiRepository(WebRequestsManager webRequestsManager) {
        this.webRequestsManager = webRequestsManager;
    }

    public Observable<List<WordTranslation>> getList() {
        return webRequestsManager.getAllWords()
                // .filter(wordTranslations -> !wordTranslations.isEmpty())
                .toObservable();
    }

    //@Override
    public Single<WordTranslation> add(WordTranslation wordTranslation) {
        return webRequestsManager.insertWord(wordTranslation);
    }

    @Override
    public Single<List<EntityIdentificator>> addAll(List<WordTranslation> wordTranslations) {
        return webRequestsManager.insertAllWords(wordTranslations);
    }

    @Override
    public Completable removeAllServerIds(List<Integer> serverIds) {
        return Single.fromCallable(() -> WordsUtilKt.absList(serverIds))
                .flatMap(serverIds1 -> webRequestsManager.removeAllServerIds(serverIds1))
                .ignoreElement();
    }

    @Override
    public Single<PullWordsAnswer> pullWords(List<Integer> wordTranslations) {
        return webRequestsManager.pullWords(wordTranslations);
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/translation/data/repository/TranslationWordsWebApiRepository.kt
package ru.inwords.inwords.translation.data.repository

import io.reactivex.Completable
import io.reactivex.Single
import ru.inwords.inwords.core.deferred_entry_manager.EntityIdentificator
import ru.inwords.inwords.network.AuthorisedRequestsManager
import ru.inwords.inwords.translation.converter.LookupReplyConverter
import ru.inwords.inwords.translation.converter.TrainingPairConverter
import ru.inwords.inwords.translation.converter.WordTranslationReplyConverter
import ru.inwords.inwords.translation.data.grpc.DictionaryGrpcService
import ru.inwords.inwords.translation.domain.model.Definition
import ru.inwords.inwords.translation.domain.model.LookupDirection
import ru.inwords.inwords.translation.domain.model.LookupDirection.EN_RU
import ru.inwords.inwords.translation.domain.model.LookupDirection.RU_EN
import ru.inwords.inwords.translation.domain.model.PullWordsAnswer
import ru.inwords.inwords.translation.domain.model.WordTranslation
import kotlin.math.abs

class TranslationWordsWebApiRepository internal constructor(
    private val dictionaryGrpcService: DictionaryGrpcService,
    private val authorisedRequestsManager: AuthorisedRequestsManager
) : TranslationWordsRemoteRepository {

    private val wordTranslationReplyConverter = WordTranslationReplyConverter()
    private val trainingPairConverter = TrainingPairConverter()
    private val lookupReplyConverter = LookupReplyConverter()

    override fun insertAllWords(wordTranslations: List<WordTranslation>): Single<List<EntityIdentificator>> {
        return authorisedRequestsManager.wrapRequest(dictionaryGrpcService.addWords(wordTranslations))
            .map { list -> list.wordIdsList.map { EntityIdentificator(it.localId.toLong(), it.serverId) } }
    }

    override fun removeAllByServerId(serverIds: List<Int>): Completable {
        return Single.fromCallable { absList(serverIds) }
            .flatMapCompletable { authorisedRequestsManager.wrapRequest(dictionaryGrpcService.deleteWords(it)) }
    }

    override fun pullWords(wordTranslations: List<Int>): Single<PullWordsAnswer> {
        return authorisedRequestsManager.wrapRequest(dictionaryGrpcService.pullWords(wordTranslations))
            .map { PullWordsAnswer(it.toDeleteList, wordTranslationReplyConverter.convertList(it.toAddList)) }
    }

    override fun lookup(text: String, lookupDirection: LookupDirection): Single<List<Definition>> {
        val lang = when (lookupDirection) {
            EN_RU -> "en-ru"
            RU_EN -> "ru-en"
        }

        return authorisedRequestsManager.wrapRequest(dictionaryGrpcService.lookup(text, lang))
            .map { lookupReplyConverter.convert(it) }
    }

    override fun trainingWords(): Single<List<WordTranslation>> {
        return authorisedRequestsManager.wrapRequest(dictionaryGrpcService.trainingWords())
            .map { reply -> reply.pairsList.map { trainingPairConverter.convert(it) } }
    }

    override fun trainingIds(): Single<List<Int>> {
        return authorisedRequestsManager.wrapRequest(dictionaryGrpcService.trainingIds())
            .map { it.userWordPairsList }
    }

    private fun absList(integers: List<Int>): List<Int> = integers.map {
        if (it < 0) {
            Unit //TODO for debug
        }
        abs(it)
    }
}
